packages/backend/src/open_payments/auth/middleware.ts

```diff
@@ -90,41 +114,32 @@ export function createTokenIntrospectionMiddleware({
         throw new OpenPaymentsServerRouteError(403, 'Inactive Token')
       }
 
-      // TODO
-      // https://github.com/interledger/rafiki/issues/835
-      const access = tokenInfo.access.find((access: Access) => {
-        if (
-          access.type !== requestType ||
-          (access.identifier && access.identifier !== ctx.walletAddressUrl)
-        ) {
-          return false
-        }
-        if (
-          requestAction === AccessAction.Read &&
-          access.actions.includes(AccessAction.ReadAll)
-        ) {
-          ctx.accessAction = AccessAction.ReadAll
-          return true
-        }
-        if (
-          requestAction === AccessAction.List &&
-          access.actions.includes(AccessAction.ListAll)
-        ) {
-          ctx.accessAction = AccessAction.ListAll
-          return true
-        }
-        return access.actions.find((tokenAction: AccessAction) => {
-          if (isActiveTokenInfo(tokenInfo) && tokenAction === requestAction) {
-            ctx.accessAction = requestAction
-            return true
-          }
-          return false
-        })
+      const access = validateTokenAccess(tokenInfo, {
+        type: requestType,
+        action: requestAction,
+        identifier: ctx.walletAddressUrl
       })
-
       if (!access) {
-        throw new OpenPaymentsServerRouteError(403, 'Insufficient Grant')
+        throw new OpenPaymentsServerRouteError(
+          403,
+          'Token info access does not match request access'
+        )
       }
+
+      if (
+        requestAction === AccessAction.Read &&
+        (access as Access).actions.includes(AccessAction.ReadAll)
+      ) {
+        ctx.accessAction = AccessAction.ReadAll
+      } else if (
+        requestAction === AccessAction.List &&
+        (access as Access).actions.includes(AccessAction.ListAll)
+      ) {
+        ctx.accessAction = AccessAction.ListAll
+      } else {
+        ctx.accessAction = requestAction
+      }
+
```

Comment:
Could we de-duplicate this logic some by returning RequestAction | undefined from `validateTokenAccess` instead of the `AccessItem`?

Looks like the only other place we use the `AccessItem` is to set the grant limits but I think the `tokenInfo` should work fine for that instead, if we wanted.

Reply:
er... maybe not. I didn't fully understand the type of `tokenInfo`. Dont think we can use that for grant limit since we won't know which access item we're dealing with. Guess we could return both access and accessAction from `validateToken` if we cared to.

Reply:
Looks like we can just return a filtered version of `access` with a single item from the introspection endpoint, and have the RS just assume that the lone item in that array is the relevant access.

Then the backend can just check if there is a `Read-All` or `List-All` action in the list if the request for a `Read` or `List` action, respectively.

[From the Resource Server Spec (emphasis mine)](https://datatracker.ietf.org/doc/html/draft-ietf-gnap-resource-servers-05#name-token-introspection):
> access (array of strings/objects):
REQUIRED. The access rights associated with this access token. This MUST be in the format described in the [Section 8](https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol-18#section-8) of [[GNAP](https://datatracker.ietf.org/doc/html/draft-ietf-gnap-core-protocol-18)]. _**This array MAY be filtered or otherwise limited for consumption by the identified RS,**_ including being an empty array, indicating that the token has no explicit access rights that can be disclosed to the RS.